docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Values, objects, and expressions](#values-objects-and-expressions)
+    -   [Expression categories](#expression-categories)
+-   [Binding patterns and local variables with `let` and `var`](#binding-patterns-and-local-variables-with-let-and-var)
+    -   [Local variables](#local-variables)
+    -   [Consuming function parameters](#consuming-function-parameters)
+-   [Reference expressions](#reference-expressions)
+    -   [Durable reference expressions](#durable-reference-expressions)
+    -   [Ephemeral reference expressions](#ephemeral-reference-expressions)
+-   [Value expressions](#value-expressions)
+    -   [Comparison to C++ parameters](#comparison-to-c-parameters)
+    -   [Value representation and customization](#value-representation-and-customization)
+    -   [Polymorphic types](#polymorphic-types)
+    -   [Interop with C++ `const &` and `const` methods.](#interop-with-c-const--and-const-methods)
+    -   [Escape hatches for value addresses in Carbon](#escape-hatches-for-value-addresses-in-carbon)
+-   [Initializing expressions](#initializing-expressions)
+    -   [Function calls and returns](#function-calls-and-returns)
+        -   [Deferred initialization from values and references](#deferred-initialization-from-values-and-references)
+        -   [Declared `returned` variable](#declared-returned-variable)
+-   [Pointers](#pointers)
+    -   [Reference types](#reference-types)
+    -   [Pointer syntax](#pointer-syntax)
+    -   [Syntax-free dereference and address-of](#syntax-free-dereference-and-address-of)
+    -   [Dereferencing customization](#dereferencing-customization)
+-   [`const`-qualified types](#const-qualified-types)
+-   [Lifetime overloading](#lifetime-overloading)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Values, objects, and expressions
+
+Carbon has both abstract _values_ and concrete _objects_. Carbon _values_ are
+things like `42`, `true`, and `i32` (a type value). Carbon _objects_ have
+_storage_ where values can be read and written. Storage also allows taking the
+address of an object in memory in Carbon.
+
+Both objects and values can be nested within each other. For example
+`(true, true)` is both a value and also contains two sub-values. When a
+two-tuple is stored somewhere, it is both a tuple-typed object and contains two
+subobjects.
+
+These terms are important components in the describing the semantics of Carbon
+code, but they aren't sufficient. We also need to explicitly and precisely talk
+about the Carbon _expressions_ that produce or reference values and objects.
+Categorizing the expressions themselves allows us to be more precise and
+articulate important differences not captured without looking at the expression
+itself.
+
+### Expression categories
```

Comment:
This PR should probably also update the "Value categories and value phases" section of `docs/design/README.md`